plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hadoop/TrinoHadoopCatalog.java

```diff
@@ -32,202 +40,227 @@
-    private final Cache<String, Catalog> catalogCache;
+    private static final Joiner SLASH = Joiner.on("/");
+    private static final Splitter DOT = Splitter.on('.');
+    private static final int PER_QUERY_CACHES_SIZE = 1000;
+    private final TrinoFileSystem trinoFileSystem;
+    private final CatalogName catalogName;
     private final String warehouse;
+    private final IcebergTableOperationsProvider tableOperationsProvider;
+    private final Cache<SchemaTableName, TableMetadata> tableMetadataCache = EvictableCacheBuilder.newBuilder()
+            .maximumSize(PER_QUERY_CACHES_SIZE)
+            .build();
 
-    public TrinoHadoopCatalog(CatalogName catalogName, HdfsEnvironment hdfsEnvironment, TypeManager typeManager, IcebergTableOperationsProvider tableOperationsProvider, boolean useUniqueTableLocation, IcebergConfig config)
+    public TrinoHadoopCatalog(
+            CatalogName catalogName,
+            TypeManager typeManager,
+            ConnectorIdentity identity,
+            IcebergTableOperationsProvider tableOperationsProvider,
+            TrinoFileSystemFactory fileSystemFactory,
+            boolean useUniqueTableLocation,
+            IcebergConfig icebergConfig)
     {
-        super(catalogName, typeManager, tableOperationsProvider, useUniqueTableLocation);
-        this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
-        this.catalogProperties = convertToCatalogProperties(config);
-        this.catalogCache = SafeCaches.buildNonEvictableCache(CacheBuilder.newBuilder().maximumSize(config.getCatalogCacheSize()));
-        this.warehouse = requireNonNull(config.getCatalogWarehouse(), "warehouse is null");
-    }
+        super(catalogName, typeManager, tableOperationsProvider, fileSystemFactory, useUniqueTableLocation);
+        this.warehouse = requireNonNull(icebergConfig.getCatalogWarehouse(), "warehouse is null");
+        this.trinoFileSystem = fileSystemFactory.create(identity);
+        this.catalogName = catalogName;
+        this.tableOperationsProvider = tableOperationsProvider;
 
-    /**
-     * generate a unique string that represents the session information.
-     * The string is used as the cache key, if the session key is the same, it means the same catalog can be reused.
-     * The default behavior is to use {@link ConnectorSession#getQueryId()} as the cache key,
-     * which means catalog is not shared across queries.
-     * Implementations can override this method to use for example the authZ user information of the session instead.
-     *
-     * @param session session
-     * @return session cache key
-     */
-    public String getSessionCacheKey(ConnectorSession session)
-    {
-        return session.getQueryId();
+        checkArgument(
+                !Strings.isNullOrEmpty(icebergConfig.getCatalogWarehouse()),
+                "Cannot initialize HadoopCatalog because warehousePath must not be null or empty");
```

Comment:
We should move this above line 118 and remove the invocation of `requireNonNull` on that line.

Reply:
Done.